src/dbnode/x/xio/types.go

```diff
@@ -37,6 +38,25 @@ type BlockReader struct {
 	BlockSize time.Duration
 }
 
+// ReadSegment reads the Segment, blocking until the segment is available or the deadline expires.
+func (b BlockReader) ReadSegment(ctx context.Context) (ts.Segment, error) {
+	done := make(chan struct{}, 1)
+	var (
+		segment ts.Segment
+		err     error
+	)
+	go func() {
+		segment, err = b.Segment()
+		done <- struct{}{}
+	}()
```

Comment:
This will spin up a goroutine per-block which feels very heavy.

Given that we do this per block, might it be ok enough to just check the cancellation on entry to the function? At worst you'll wait for one block longer to be retrieved and all the block reads will be cancelled properly before being actually read (which actually protects the DB more than this goroutine still being alive).

i.e.
```go
	select {
	case <-ctx.GoContext().Done():
		return ts.Segment{}, ctx.GoContext().Err()	
	default:
	}

	return b.Segment()
```